add some! macro for unwrapping Option more safely

vitiral opened the pull request:
This is an RFC for adding the `some!` macro to unwrap `Option` types.

Ticki commented:
Why not just make `try!` generic, by adding a trait for optional types (Result, Option, etc.)?

vitiral commented:
I included that as an alternative that I thought of later. That would certainly be preferable in my view, as it mimicks `unwrap`. In fact, as I was learning the language, I tried to use `try!` repeatedly before finding out it does not support `Option` types, so this would also be better for new users.

All of the arguments in this RFC still apply to `try`, without hardly any of the negatives. I can re-write the RFC for try if that is what people want.

vitiral commented:
@Ticki how would you implement that?

Ticki commented:
By implementing `From<Result<T, U>>` for `Option<T>`.

vitiral commented:
Ah, you are saying it is possible to make the existing `try!` macro work for any type. That is pretty cool! The only thing I would be worried about is whether that affects performance -- but I would hope the compiler would optimize all those type conversions away.

Also, would this require using the `Into<Option<T>>` (like in #1402) any time you wanted to use `try!` in your function? Sorry if that is a dumb question, I am new to rust and don't understand why implementing the `From` trait there requires being more verbose in the function signatures.

Ticki commented:
> Also, would this require using the Into<Option<T>> (like in #1402) any time you wanted to use try! in your function

Nope. The conversion is handled by the macro.

> Sorry if that is a dumb question

It's not!

jonas-schievink commented:
:-1: `Option` already has 21 methods, not counting trait impls. I'm using Rust since over a year and still don't know when to use which one. No need to make it even more complicated.

Ticki commented:
@jonas-schievink That is why we could integrate this into the `try!` macro.

main-- commented:
I'm having trouble coming up with usecases where this is both the best and a particularly ergonomic solution. `try!()` is very useful not only because of the safe unwrap it provides but also mainly because it *propagates errors*. It's the perfect tool for when you need to just pass an error down the call stack because you have no straightforward way to immediately handle it. Note how `Option<T>` is semantically equivalent to `Result<T, ()>`. It's still good to have two distinct types because they convey two different meanings (that are intermixed in the classical `null` other languages have): `None` simply signals the absence of a value while `Err` clearly shows that an error occurred.

**`try!()`'s error propagation semantics are not generally desirable with `Option<T>`-returning APIs**.

This is quite a controversial claim, but I'd say it's obvious once you think about it: `try!()` only really shines when you use it many times within a function. Otherwise, a quick `match`/`if let Ok(x)`/`and_then` does the job just fine. Again, `try!()` merely pushes the error down to the caller. At some point, we still have to handle it *somehow*. And this is the part that gets tricky under this proposal: *There are no errors in the `Option<T>` world*. Once you're 10 frames lower in the callstack, you have absolutely no idea where this `None` came from. All you know is that *something* went wrong. What you really want in all of these cases is not `some!(my_option)` but [`try!(my_option.ok_or(TheError))`](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.ok_or). This is a quick and elegant way to convert your `Option<T>` into a `Result<T, E>` and then benefit from all the infrastructure we already have there (`try!()` and maybe one day also the `?` operator).

It appears to be the general consensus (and I fully agree) that `unwrap()` is generally problematic and we should make it very clear especially to programmers new to Rust that this is very often not the right thing to do. But consider: Even though it's not directly obvious, the `some!()` macro proposed here is actually *worse* than `unwrap()` as the panic caused by the latter at least prints a stacktrace.

* `unwrap()`: Your program crashes but you know why and can fix the bug.
* `some!()`: Your program terminates gracefully but you have absolutely no idea why.

---

Of course I may have missed a scenario where my criticism doesn't apply. But unless we come up with one, I don't think we should implement this RFC.

vitiral commented:
Note: I would like to formally change this RFC to be recommending extending `try!` rather than adding `option!`

@main-- all of those are very compelling arguments, and certainly good reason to keep the current behavior. However, there are several reasons I think this functionality would be good and I want to make their case.

As explained in the RFC, I come at this from two perspectives:
- `unwrap` is used extensively in documentation "for the sake of brevity". However, demonstrating bad coding practices for the sake of brevity, when an alternative *could* be available that uses much better coding practices, is not a good idea in my opinion. You're examples of using `ok_or` with `try` would be much better coding practice for a large majority of them.
- splitting up logic from a function.

Let's look at the second case, since I think it most addresses your concerns. From my experience doing data analysis, I can think of several cases where I was doing a particularily large amount of logic, and was splitting it into member functions. Because rust is a very safe language that requires you to address each posibility individually, it is totally reasonable (and possible) for every helper function branch to validly return `None` (but not `Err`), because you know that it errored for a valid reason and you will always choose to just skip it if it returns None. Furthermore, it is totally possible for a single function to use several `try!` macros in order to shortcut the `unwrap` of `Option`.

The trivial case in the pull request (#30951) I gave highlights it (I have modified it to use `try!` instead of `some!`):
```
fn double_divide(num: f64, den: f64) -> Option<f64> {
    let value = try!(divide(num, den));
    Some(value * 2f64)
}

fn double_divide(num: f64, den: f64) -> Option<f64> {
    let value = match divide(num, den) {
        Some(value) => value,
        None => return None,
    };
    Some(value * 2f64)
}
```

This is a totally reasonable "propogation of None" -- no information is being lost, it is sucinct and ends up with the same api as the `divide` function it is calling, which is completely reasonable since it has exactly the same error.

Thankyou for taking the time to write your response. It was very informative and is something that should be highlighted in documentation over the choice of None vs Err. In particular, the following is my impression of that choice:
- return `Err` when the problem is not expected, needs description, or there is more than one possible error that could be handled different ways.
- return `None` when the potential failure case is obvious and simple to handle, i.e. when a number doesn't divide, when an algorithm can only fail because of an obvious reason (i.e. takes an iterator and *can only* returns None if there are no items in the iterator), etc.

Whether that the first and second reasons combined are succicient to include this in the language, I will leave to more experienced Rustians than me. Thanks for reading!

vitiral commented:
So, to summarize: the main uses cases are:
- use cases where you have a particularily large amount of logic that cannot have an error except for one obvious case. In these cases, it is good programming practice to split apart your logic, which might require chaining explicity unwraps of Some types
- use cases where you are extending an existing API that retuns None, such as the example with `double_divide`.

main-- commented:
So the way I understand the first case right now is: You have an existing function that performs early returns all the time and you'd like to split it up. Therefore, the early return translates into returning `None` to the caller who then actually bails out.

Regarding the second case, I'd say there is little room for misunderstandings. :)

There are numerous ways to solve the example you outlined, but the most natural one in this particular case is `map`:

```rust
fn double_divide3(num: f64, den: f64) -> Option<f64> {
    divide(num, den).map(|x| x * 2f64)
}
```

I wrote a few quick [variations](http://is.gd/Vc9iUM) on that and I'm still convinced that `map` and [`and_then`](https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.and_then) both do a great job at covering most - if not all - usecases.

That being said, there is the [try_opt](https://crates.io/crates/try_opt/) crate which provides what's suggested here (except the `try!` integration). There's also #1394 which is (ignoring this final change) pretty much identical.

Ticki commented:
To do `if let ...` is overly verbose for doing such little thing.

vitiral commented:
I confess I did not realize that `Option.map` or `if let` were so useful in these contexts. I am obviously way too new to the language. I still contend that differing `try` behavior is confusing for newbies, but that is not an insurmountable challenge, and if it is believed that this change will actually *encourage* bad coding practices, then I would say scrap it.

Thank you for the information and discussion.

tbu- commented:
Also present in this crate: [`unwrap_or_return!` in `mac`](https://simonsapin.github.io/kuchiki/mac/macro.unwrap_or_return!.html)

alexcrichton commented:
I believe this RFC is almost the same as https://github.com/rust-lang/rfcs/pull/1394, perhaps they could be merged?

vitiral commented:
Yes, that would be fine. Feel free to close this one.

aturon commented:
Closing in favor of #1394.
